Pad: don't error when unused fill value is zero

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #73434
* __->__ #76307

Fixes pytorch/vision#5873

In the python version of `F.pad`, checking that the fill value was
left as default was done by comparing against zero:
https://github.com/pytorch/pytorch/blob/bc2c6edaf163b1a1330e37a6e34caf8c553e4755/torch/nn/functional.py#L4366

So if someone does explicitly pass in a zero-value, then this
`TORCH_CHECK` was an accidental BC-break.

peterbell10 committed:
Pad: don't error when unused fill value is zero

Fixes pytorch/vision#5873

In the python version of `F.pad`, checking that the fill value was
left as default was done by comparing against zero:
https://github.com/pytorch/pytorch/blob/bc2c6edaf163b1a1330e37a6e34caf8c553e4755/torch/nn/functional.py#L4366

So if someone does explicitly pass in a zero-value, then this
`TORCH_CHECK` was an accidental BC-break. Instead, we should just warn
in that case.

[ghstack-poisoned]

datumbox reviewed (commented):
Thanks for the change @peterbell10. I confirm that this should resolve the BC problem and fix the downstream issues at TorchVision. 


The only question I have is on whether we should be raising a deprecation warning. Hence, I'll leave it to @albanD to approve and merge if the current approach aligns with Core's deprecation policy.

albanD commented on aten/src/ATen/native/PadNd.cpp:
```diff
@@ -163,10 +163,13 @@ Tensor _pad_enum(const Tensor &self, IntArrayRef pad, int64_t mode_int, c10::opt
+                            "\" doesn't take in value argument");
+    // Emit error if value != 0, otherwise just warn
+    TORCH_CHECK(*value == 0, err_msg);
+    TORCH_WARN(err_msg);
```
Do we really want to deprecate this? Or are we just happy with allowing `0.` here?
Given that people use this already and this is not very harmful. I would vote for not changing the current behavior. At least for now so that we can quickly land this. We can discuss (with Joel) if we want to deprecate that later.

datumbox replied:
I agree that maintaining the current behaviour until we figure out what to do is probably preferable. 

Note that we already have a PR at https://github.com/pytorch/vision/pull/5875 which mitigates the issue by not passing any parameter in specific modes but maintaining BC is worth it for other users.

peterbell10 replied:
Okay, I've removed the warning. I would argue that a passing parameter that gets ignored is somewhat confusing, but if it's already baked into torchvision then I guess it's too late to fix.

jbschlosser reviewed (commented):
Thanks for the quick forward fix!

re: Vasilis' question: Our policy in core is to throw deprecation warnings using [TORCH_WARN_ONCE](https://github.com/pytorch/pytorch/wiki/Developer-FAQ#when-should-i-torch_warn-vs-torch_warn_once) that are actionable to silence the warning. The warning here can be silenced by passing `None`  instead of `0` to torch's `pad`.

That said, it seems like torchvision's public API for `pad` would have to change to pass `None` as the default value to torch's pad. Otherwise, this warning will be thrown under normal usage with no way to remove it - is this true?

If so, maybe we should just avoid the deprecation for now to minimize pain. As @albanD mentioned, it isn't a particularly harmful thing to pass an unused fill value.

peterbell10 committed:
Remove warning for value=0 on "Pad: don't error when unused fill value is zero"


Fixes pytorch/vision#5873

In the python version of `F.pad`, checking that the fill value was
left as default was done by comparing against zero:
https://github.com/pytorch/pytorch/blob/bc2c6edaf163b1a1330e37a6e34caf8c553e4755/torch/nn/functional.py#L4366

So if someone does explicitly pass in a zero-value, then this
`TORCH_CHECK` was an accidental BC-break.

[ghstack-poisoned]

albanD reviewed (approved):
SGTM!

datumbox reviewed (approved):
LGTM, thanks!

FYI TorchVision is going to switch to passing the value only when needed (see https://github.com/pytorch/vision/pull/5875), so that we can fix ASAP our CI.

albanD commented:
@pytorchbot merge this please
